src/routes/webhooks/paddle.ts

```diff
@@ -405,6 +370,66 @@ const notifyNewPaddleTransaction = async ({
   await webhooks.transactions.send({ blocks });
 };
 
+export const processGiftedPayment = async (
+  data: TransactionCompletedEvent['data'],
+) => {
+  const con = await createOrGetConnection();
+  const { gifter_id, user_id } = data.customData as PaddleCustomData;
+
+  if (user_id === gifter_id) {
+    logger.error({ type: 'paddle' }, 'User and gifter are the same');
+    return;
+  }
+
+  const gifterUser = await con.getRepository(User).findOneBy({ id: gifter_id });
+
+  if (!gifterUser) {
+    logger.error({ type: 'paddle' }, 'Gifter user not found');
+    return;
+  }
+
+  const targetUser = await con.getRepository(User).findOne({
+    select: ['subscriptionFlags'],
+    where: { id: user_id },
+  });
+
+  if (isPlusMember(targetUser?.subscriptionFlags?.cycle)) {
+    logger.error({ type: 'paddle' }, 'User is already a Plus member');
+    return;
+  }
+
+  await con.getRepository(User).update(
+    {
+      id: user_id,
+    },
+    {
+      subscriptionFlags: updateSubscriptionFlags({
+        cycle: SubscriptionCycles.Yearly,
+        createdAt: data?.createdAt,
+        subscriptionId: data?.id,
+        gifterId: gifter_id,
+        giftExpirationDate: addMilliseconds(
+          new Date(),
+          plusGiftDuration,
+        ).toISOString(),
+      }),
+      flags: updateFlagsStatement({ showPlusGift: true }),
+    },
+  );
+};
+
+export const processTransactionCompleted = async (
+  result: TransactionCompletedEvent,
+) => {
+  const { gifter_id } = (result?.data?.customData ?? {}) as PaddleCustomData;
+
+  if (gifter_id) {
+    await processGiftedPayment(result.data);
+  }
```

Comment:
lgtm, but its a bit hard to see what you added since everything in code moved, what was the actual change?

Reply:
Yeah, most things were just moved, and this is the only new thing.

Reply:
but what inside `processGiftedPayment` is new?

Reply:
I mean this block, the content of `processGiftedPayment` was just moved from the deleted lines above.

Reply:
https://github.com/dailydotdev/daily-api/pull/2648#discussion_r1947480532

This basically. Also to make it easier to review since the whole function was just moved, and this is the new thing that might be harder to review if I had kept them in the same place.

Reply:
🤝